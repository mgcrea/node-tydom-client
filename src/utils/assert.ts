import {AssertionError} from 'assert';

export function assert(condition: unknown, message?: string): asserts condition {
  if (!condition) {
    throw new AssertionError({message});
  }
}
