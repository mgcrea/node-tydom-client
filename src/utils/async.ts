import {RetryObject} from 'got/dist/source';

export const asyncWait = (ms: number): Promise<void> =>
  new Promise((resolve) => {
    setTimeout(resolve, ms);
  });

export const calculateDelay = ({attemptCount}: Pick<RetryObject, 'attemptCount'>): number =>
  Math.min(1000 * Math.pow(2, Math.max(1, attemptCount)) + Math.random() * 100, Math.pow(2, 31) - 1);
