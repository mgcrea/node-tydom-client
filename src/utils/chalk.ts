import chalk from 'chalk';

export const chalkString = (s: unknown) => chalk.green(`'${s}'`);
export const chalkKeyword = (s: unknown) => chalk.blue(`${s}`);
export const chalkJson = (s: unknown) => chalk.grey(JSON.stringify(s));
export const chalkStringArray = (a: Array<unknown>) => `[ ${a.map(chalkString).join(', ')} ]`;
export const chalkNumber = (n: unknown) => chalk.yellow(`${n}`);
export const chalkBoolean = (b: unknown) => chalk.yellow(b ? 'true' : 'false');
export const chalkDate = (d: Date) => chalk.magenta(d.toISOString());
