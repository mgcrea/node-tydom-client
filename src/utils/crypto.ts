import crypto from 'crypto';

export const md5 = (data: crypto.BinaryLike): Buffer =>
  crypto
    .createHash('md5')
    .update(data)
    .digest();

export const getRandomBytes = async (size = 256): Promise<Buffer> => {
  return new Promise((resolve, reject) => {
    crypto.randomBytes(size, (err, buf) => {
      if (err) {
        reject(err);
        return;
      }
      resolve(buf);
    });
  });
};

let requestCounterValue = 0;
export const getRequestCounter = () => {
  requestCounterValue += 1;
  const nc = requestCounterValue + '';
  return ('00000000' + nc).substr(nc.length);
};
