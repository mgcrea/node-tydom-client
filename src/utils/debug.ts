// @ts-ignore
import {name} from './../../package.json';
import createDebug from 'debug';

const debug = createDebug(name);

export default debug;

export const dir = (...args: unknown[]) => {
  console.dir(...args, {colors: true, depth: 10});
};
