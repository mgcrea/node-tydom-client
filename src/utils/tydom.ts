import fetch from 'node-fetch';
import assert from 'assert';
import {getRequestCounter, getRandomBytes, md5} from './crypto';

export type TydomResponse = {
  status: number;
  body: string;
  headers: Map<string, string>;
  json: () => Promise<{[s: string]: string}>;
};

export const castTydomResponse = (body: string, headers: Map<string, string>): TydomResponse => {
  const hasBody = body.length > 0;
  const shouldBeJson = headers.has('content-type') && headers.get('content-type')!.includes('application/json');
  const isActuallyHtml = hasBody && body.startsWith('<!doctype html>');
  const status = isActuallyHtml ? 400 : 200;
  const json = async () => {
    if (!hasBody || !shouldBeJson) {
      return {};
    }
    if (isActuallyHtml) {
      return {error: 1};
    }
    return JSON.parse(body);
  };
  return {status, body, headers, json};
};

type GetTydomDigestAccessAuthenticationOptions = {
  username: string;
  hostname: string;
  headers?: {[s: string]: string};
};

export const getTydomDigestAccessAuthenticationFields = async ({
  username,
  hostname,
  headers
}: GetTydomDigestAccessAuthenticationOptions) => {
  const query = `mac=${username}&appli=1`;
  const path = `/mediation/client?${query}`;
  const method = 'GET';
  console.log(`Connecting to url="${`https://${hostname}${path}`}"`);
  const {status, headers: responseHeaders} = await fetch(`https://${hostname}${path}`, {method, headers});
  assert(status === 401, `Unexpected status=${status}`);
  const authHeader = responseHeaders.get('www-authenticate');
  assert(authHeader, 'Missing required "www-authenticate" header');
  const authFieldsSplit = (authHeader as string).replace(/^Digest\s+/, '').split(',');
  const authFields = authFieldsSplit.reduce((soFar: {[s: string]: string}, field: string) => {
    const [key, value] = field.split('="');
    soFar[key.trim()] = value.slice(0, -1);
    return soFar;
  }, {});
  return authFields as DigestAccessAuthenticationFields;
};

type DigestAccessAuthenticationOptions = {
  uri: string;
  username: string;
  password: string;
};

type DigestAccessAuthenticationFields = {
  realm: string;
  qop: string;
  nonce: string;
  opaque?: string;
};

type DigestAccessAuthenticationHeader = {
  header: string;
  response: string;
  nc: string;
  cnonce: string;
};

export const computeDigestAccessAuthenticationHeader = async (
  {uri, username, password}: DigestAccessAuthenticationOptions,
  {realm, qop, nonce /*, opaque */}: DigestAccessAuthenticationFields
): Promise<DigestAccessAuthenticationHeader> => {
  const nc = getRequestCounter();
  const cnonce = (await getRandomBytes(4)).toString('hex');
  const ha1 = (await md5(`${username}:${realm}:${password}`)).toString('hex');
  const ha2 = (await md5(`GET:${uri}`)).toString('hex');
  const res = `${ha1}:${nonce}:${nc}:${cnonce}:${qop}:${ha2}`;
  const response = (await md5(res)).toString('hex');
  const header = `Digest username="${username}", realm="${realm}", nonce="${nonce}", uri="${uri}", response="${response}", qop=${qop}, nc=${nc}, cnonce="${cnonce}"`;
  return {header, response, nc, cnonce};
};
