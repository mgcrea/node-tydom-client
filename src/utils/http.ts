import {HTTPParser} from 'http-parser-js';
import {assert} from './assert';
import {getRandomBytes, getRequestCounter, md5} from './crypto';
import {castTydomMessage, TydomBinaryMessage, TydomHttpMessage} from './tydom';

const requestParser = new HTTPParser(HTTPParser.REQUEST);
const responseParser = new HTTPParser(HTTPParser.RESPONSE);

const REQUEST_REGEX = /^([A-Z-]+) ([^ ]+) HTTP\/(\d)\.(\d)$/;
const RESPONSE_REGEX = /^HTTP\/(\d)\.(\d) (\d{3}) ?(.*)$/;

export type MessageType = 'response' | 'request' | 'binary';
export const getMessageType = (data: Buffer): {type: MessageType; matches: RegExpMatchArray | null} => {
  const firstLine = data.slice(0, data.indexOf('\r\n')).toString('ascii');
  if (RESPONSE_REGEX.test(firstLine)) {
    return {type: 'response', matches: firstLine.match(RESPONSE_REGEX)};
  }
  if (REQUEST_REGEX.test(firstLine)) {
    return {type: 'request', matches: firstLine.match(REQUEST_REGEX)};
  }
  return {type: 'binary', matches: null};
};

export const parseIncomingMessage = async (data: Buffer): Promise<TydomBinaryMessage | TydomHttpMessage> => {
  return new Promise((resolve, reject) => {
    try {
      const {type: messageType, matches: messageMatches} = getMessageType(data);
      if (messageType === 'binary') {
        resolve({type: messageType, data});
        return;
      }
      assert(messageMatches, 'Unexpected empty messageMatches');
      const parser = messageType === 'response' ? responseParser : requestParser;
      const bodyParts: Buffer[] = [];
      let headers = new Map<string, string>();
      parser.onHeadersComplete = (res: {headers: string[]}) => {
        headers = res.headers.reduce<{headers: Map<string, string>; lastValue: string}>(
          (soFar, value, index) => {
            index % 2 ? soFar.headers.set(soFar.lastValue.toLowerCase(), value) : (soFar.lastValue = value);
            return soFar;
          },
          {headers, lastValue: ''},
        ).headers;
      };
      parser.onBody = (chunk: Buffer, offset: number, length: number) => {
        bodyParts.push(chunk.slice(offset, offset + length));
      };
      parser.onMessageComplete = () => {
        const date = new Date();
        const body = Buffer.concat(bodyParts).toString('utf8');
        switch (messageType) {
          case 'response': {
            const method = null;
            const uri = headers.get('uri-origin') || '/';
            const status = parseInt(messageMatches[3], 10);
            resolve(castTydomMessage({type: messageType, method, uri, status, body, headers, date}));
            return;
          }
          case 'request': {
            const method = messageMatches[1];
            const uri = messageMatches[2];
            const status = null;
            resolve(castTydomMessage({type: messageType, method, uri, status, body, headers, date}));
            return;
          }
          default: {
            throw new Error(`Unhandled messageType=${messageType}`);
          }
        }
      };
      parser.execute(data);
    } catch (err) {
      reject(err);
    }
  });
};

export type BuildRawHttpRequestOptions = {
  url: string;
  method: 'GET' | 'PUT' | 'POST' | 'DELETE';
  headers?: {[s: string]: string};
  body?: string;
};

export const buildRawHttpRequest = ({url, method, headers = {}, body}: BuildRawHttpRequestOptions): string => {
  const rawRequest = `${method} ${url} HTTP/1.1`;
  const rawHeaders = Object.keys(headers).reduce((soFar, key) => {
    return `${soFar}${key}: ${headers[key]}\r\n`;
  }, '');
  return `${rawRequest}\r\n${rawHeaders}\r\n\r\n${body ? `${body}\r\n\r\n` : ''}`;
};

export type DigestAccessAuthenticationOptions = {
  username: string;
  password: string;
};

export type DigestAccessAuthenticationFields = {
  realm: string;
  qop: string;
  nonce: string;
  opaque?: string;
  uri: string;
};

export type DigestAccessAuthenticationHeader = {
  header: string;
  response: string;
  nc: string;
  cnonce: string;
};

export const computeDigestAccessAuthenticationHeader = async (
  {username, password}: DigestAccessAuthenticationOptions,
  {uri, realm, qop, nonce /*, opaque */}: DigestAccessAuthenticationFields,
): Promise<DigestAccessAuthenticationHeader> => {
  const nc = getRequestCounter();
  const cnonce = (await getRandomBytes(4)).toString('hex');
  const ha1 = (await md5(`${username}:${realm}:${password}`)).toString('hex');
  const ha2 = (await md5(`GET:${uri}`)).toString('hex');
  const res = `${ha1}:${nonce}:${nc}:${cnonce}:${qop}:${ha2}`;
  const response = (await md5(res)).toString('hex');
  const header = `Digest username="${username}", realm="${realm}", nonce="${nonce}", uri="${uri}", response="${response}", qop=${qop}, nc=${nc}, cnonce="${cnonce}"`;
  return {header, response, nc, cnonce};
};
