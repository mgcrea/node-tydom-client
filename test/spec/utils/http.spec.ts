import {parseIncomingMessage} from 'src/utils/http';
import {getInfoResponse, updatedDevicesData} from '../__fixtures__/messages';
import {omit} from 'lodash';

describe('parseIncomingMessage', () => {
  it('should properly parse a getInfoResponse response', async () => {
    const message = await parseIncomingMessage(getInfoResponse);
    expect(message).toBeDefined();
    expect(Object.keys(message)).toMatchSnapshot();
    expect(omit(message, 'date')).toMatchSnapshot();
  });
  it('should properly parse a updatedDevicesData response', async () => {
    const message = await parseIncomingMessage(updatedDevicesData);
    expect(message).toBeDefined();
    expect(Object.keys(message)).toMatchSnapshot();
    expect(omit(message, 'date')).toMatchSnapshot();
  });
});
