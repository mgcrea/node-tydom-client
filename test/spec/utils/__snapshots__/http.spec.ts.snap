// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`parseIncomingMessage should properly parse a getInfoResponse response 1`] = `
Array [
  "type",
  "uri",
  "method",
  "status",
  "body",
  "headers",
  "date",
]
`;

exports[`parseIncomingMessage should properly parse a getInfoResponse response 2`] = `
Object {
  "body": Object {
    "TYDOM.dat": 1,
    "apiMode": true,
    "bdd.json": 11,
    "bddEmpty": false,
    "bddStatus": 0,
    "bdd_mig.json": 0,
    "bootReference": "P21260012",
    "bootVersion": "01.00.03",
    "clock": Object {
      "clock": "2020-01-17T13:12:13+01:00",
      "summerOffset": "ON",
      "timezone": 60,
    },
    "collect.json": 5,
    "config": "prod",
    "config.json": 1,
    "gateway.dat": 0,
    "geoloc": Object {
      "latitude": 48897412,
      "longitude": 2125472,
    },
    "groups.json": 6,
    "info_mig.json": 0,
    "keyReference": "21260011",
    "keyVersionHW": "00.00.01",
    "keyVersionSW": "01.03.03",
    "keyVersionStack": "01.01.03",
    "mac": "001A2502951F",
    "mainId": "6414118",
    "mainReference": "21260010",
    "mainVersionHW": "00.00.01",
    "mainVersionSW": "02.02.53",
    "mom.json": 0,
    "mom_api.json": 105,
    "moments": Object {
      "suspend": Object {
        "to": 0,
      },
    },
    "passwordEmpty": false,
    "productName": "TYDOM2",
    "scenario.json": 5,
    "site.json": 149,
    "updateAvailable": false,
    "urlMediation": "mediation.tydom.com",
  },
  "headers": Map {
    "server" => "Tydom-001A2502951F",
    "uri-origin" => "/info",
    "content-type" => "application/json",
    "transfer-encoding" => "chunked",
    "transac-id" => "request_1",
  },
  "method": null,
  "status": 200,
  "type": "response",
  "uri": "/info",
}
`;

exports[`parseIncomingMessage should properly parse a updatedDevicesData response 1`] = `
Array [
  "type",
  "uri",
  "method",
  "status",
  "body",
  "headers",
  "date",
]
`;

exports[`parseIncomingMessage should properly parse a updatedDevicesData response 2`] = `
Object {
  "body": Array [
    Object {
      "endpoints": Array [
        Object {
          "data": Array [
            Object {
              "name": "level",
              "validity": "upToDate",
              "value": 100,
            },
            Object {
              "name": "onFavPos",
              "validity": "upToDate",
              "value": false,
            },
          ],
          "error": 0,
          "id": 1531745761,
        },
      ],
      "id": 1531745761,
    },
  ],
  "headers": Map {
    "server" => "Tydom-001A2502951F",
    "content-type" => "application/json",
    "transfer-encoding" => "chunked",
  },
  "method": "PUT",
  "status": null,
  "type": "request",
  "uri": "/devices/data",
}
`;
